core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLRecordReader.java

```diff
@@ -0,0 +1,305 @@
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.lib.input.FileSplit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.text.DecimalFormat;
+import java.util.HashMap;
+import java.util.Map;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+
+/**
+ * XMLRecordReader class to read through a given xml document and to output xml blocks as per node path specified.
+ */
+public class XMLRecordReader extends RecordReader<LongWritable, Map<String, String>> {
+  private static final Logger LOG = LoggerFactory.getLogger(XMLRecordReader.class);
+
+  public static final String CLOSING_END_TAG_DELIMITER = ">";
+  public static final String OPENING_END_TAG_DELIMITER = "</";
+  public static final String CLOSING_START_TAG_DELIMITER = ">";
+  public static final String OPENING_START_TAG_DELIMITER = "<";
+
+  private LongWritable currentKey;
+  private Map<String, String> currentValue;
+  private String nodePath;
+  private final String fileName;
+  private final XMLStreamReader reader;
+  private final String[] nodes;
+  private final int totalNodes;
+  private final Map<Integer, String> actualNodeLevelMap;
+  private Map<Integer, String> currentNodeLevelMap;
+  private int nodeLevel = 0;
+  private String tempFilePath = null;
+  private final Path file;
+  private String fileAction;
+  private FileSystem fs;
+  private final String targetFolder;
+  private long availableBytes = 0;
+  private final PublishingInputStream inputStream;
+  private final DecimalFormat df = new DecimalFormat("#.##");
+
+  public XMLRecordReader(FileSplit split, Configuration conf) throws IOException {
+    file = split.getPath();
+    fileName = file.toUri().toString();
+    fs = file.getFileSystem(conf);
+    XMLInputFactory factory = XMLInputFactory.newInstance();
+    FSDataInputStream fdDataInputStream = fs.open(file);
+    inputStream = new PublishingInputStream(fdDataInputStream);
+    availableBytes = inputStream.available();
+    try {
+      reader = factory.createXMLStreamReader(inputStream);
+    } catch (XMLStreamException exception) {
+      throw new RuntimeException("XMLStreamException exception : ", exception);
+    }
+    //set required node path details.
+    nodePath = conf.get(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH);
+    //remove preceding '/' in node path to avoid first unwanted element after split('/')
+    if (nodePath.indexOf("/") == 0) {
+      nodePath = nodePath.substring(1, nodePath.length());
+    }
+    nodes = nodePath.split("/");
+    totalNodes = nodes.length;
+
+    //map to store node path specified nodes with key as node level
```

Comment:
what does this mean? Can "actualNodeLevelMap" be named better? "actual" doesn't really describe what it does, and its already clear that it is a map.